pkg/server/plugin/upstreamca/spire/spire_node.go

```diff
@@ -0,0 +1,135 @@
+package spireplugin
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/spiffe/spire/pkg/common/bundleutil"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/api/node"
+)
+
+func (m *spirePlugin) submitCSRUpstreamCA(ctx context.Context, nodeClient node.NodeClient, csr []byte) ([]*x509.Certificate, *bundleutil.Bundle, error) {
+	nodeRequest := &node.FetchX509SVIDRequest{
+		Csrs: [][]byte{csr},
+	}
+
+	stream, err := nodeClient.FetchX509SVID(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	err = stream.Send(nodeRequest)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	stream.CloseSend()
+
+	nodeResponse, err := stream.Recv()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return m.getCertFromResponse(nodeResponse)
+}
+
+func (m *spirePlugin) getCertFromResponse(response *node.FetchX509SVIDResponse) ([]*x509.Certificate, *bundleutil.Bundle, error) {
+	if response.SvidUpdate == nil {
+		return nil, nil, errors.New("response missing svid update")
+	}
+	if len(response.SvidUpdate.Svids) < 1 {
+		return nil, nil, errors.New("no svid received")
+	}
+
+	svidMsg, ok := response.SvidUpdate.Svids[m.trustDomain.String()]
+	if !ok {
+		return nil, nil, fmt.Errorf("no svid for %v received", m.trustDomain.String())
+	}
+
+	svid, err := x509.ParseCertificates(svidMsg.CertChain)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid svid: %v", err)
+	}
+
+	if response.SvidUpdate.Bundles == nil {
+		return nil, nil, errors.New("missing bundles")
+	}
+
+	bundleProto := response.SvidUpdate.Bundles[m.trustDomain.String()]
+	if bundleProto == nil {
+		return nil, nil, errors.New("missing bundle")
+	}
+
+	bundle, err := bundleutil.BundleFromProto(bundleProto)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid bundle: %v", err)
+	}
+
+	return svid, bundle, nil
+}
+
+func (m *spirePlugin) newNodeClientConn(ctx context.Context, wCert []byte, wKey []byte, wBundle []byte) (*grpc.ClientConn, error) {
+	conn, err := m.dialNodeAPI(ctx, wCert, wKey, wBundle)
```

Comment:
This could be
```go
return m.dialNodeAPI(ctx, wCert, wKey, wBundle)
```

Which makes me wonder if this helper method should just go away altogether and `SubmitCSR` could call `m.dialNodeAPI` directly

Reply:
Done.